File: leetcode/medium/DailyTemperatures.py
from typing import List

class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        """
        Given an array of integers `temperatures` represents the daily temperatures,
        return an array `answer` such that `answer[i]` is the number of days you
        have to wait after the i-th day to get a warmer temperature. If there is
        no future day for which this is possible, keep `answer[i] == 0` instead.

        This problem can be efficiently solved using a monotonic stack.
        The stack will store indices of temperatures in decreasing order.

        Args:
            temperatures: A list of integers representing daily temperatures.

        Returns:
            A list of integers where `answer[i]` is the number of days to wait
            for a warmer temperature, or 0 if no warmer day exists.
        """
        
        # Initialize an empty stack. The stack will store tuples of (index, temperature).
        # We maintain a decreasing order of temperatures in the stack.
        stack = []
        
        # Initialize the result array with zeros. This will store the waiting days.
        # If no warmer temperature is found, the value remains 0.
        result = [0] * len(temperatures)
        
        # Iterate through the temperatures with their indices.
        for index, temperature in enumerate(temperatures):
            # While the stack is not empty AND the current temperature is warmer
            # than the temperature at the index on top of the stack:
            while stack and stack[-1][1] < temperature:
                # Pop the (index, temperature) pair from the stack.
                # This means we found a warmer temperature for the popped day.
                prev_index, prev_temperature = stack.pop()
                
                # Calculate the number of days to wait and store it in the result array.
                result[prev_index] = index - prev_index
            
            # Push the current (index, temperature) onto the stack.
            # The stack maintains a decreasing order of temperatures.
            stack.append((index, temperature))
            
        return result

File: leetcode/medium/DailyTemperatures.java
import java.util.Deque;
import java.util.ArrayDeque;

/**
 * Given an array of daily temperatures, returns an array where each element
 * is the number of days you have to wait after the current day to get a
 * warmer temperature. If there is no future day with a warmer temperature,
 * the result is 0 for that day.
 */
class Solution {
    /**
     * Calculates the number of days to wait for a warmer temperature using a monotonic stack.
     * The stack stores the indices of days in strictly decreasing order of temperature.
     *
     * @param temperatures An array representing the daily temperatures.
     * @return An array where ans[i] is the number of days until a warmer temperature.
     */
    public int[] dailyTemperatures(int[] temperatures) {
        int n = temperatures.length;
        // The result array, initialized to 0s (default for int), which is correct for days with no warmer future.
        int[] ans = new int[n];

        // Monotonic decreasing stack. Stores the INDEX of the day.
        // The temperature for the index 'i' can be retrieved using temperatures[i].
        // Using Integer for indices simplifies the stack operations compared to int[].
        Deque<Integer> stack = new ArrayDeque<>();

        // Iterate through each day, from left to right.
        for (int i = 0; i < n; i++) {
            int currentTemperature = temperatures[i];

            // Maintain the monotonic decreasing property:
            // While the stack is not empty AND the temperature at the index on top of the stack
            // is LESS than the current day's temperature (temperatures[i]), we've found the
            // *first* warmer day for the day at the stack's top.
            while (!stack.isEmpty() && temperatures[stack.peek()] < currentTemperature) {
                // Pop the index of the day for which we've found the answer.
                int previousIndex = stack.pop();

                // The wait time is the difference between the current index (i) and the previous index.
                ans[previousIndex] = i - previousIndex;
            }

            // Push the current day's index onto the stack.
            // It potentially needs a warmer day later in the future.
            stack.push(i);
        }

        // Days remaining in the stack have no warmer day to their right, so their
        // corresponding 'ans' value remains 0 (the default value), which is correct.
        return ans;
    }
}